I did not actually check for all the type conversions, MSVC fails at:
```
D:\a\zlib-ng\zlib-ng\inffast_tpl.h(178): warning C4242: '-=': conversion from 'unsigned int' to 'uint8_t', possible loss of data [D:\a\zlib-ng\zlib-ng\zlib-ng.vcxproj]
D:\a\zlib-ng\zlib-ng\inffast_tpl.h(198): warning C4242: '-=': conversion from 'unsigned int' to 'uint8_t', possible loss of data [D:\a\zlib-ng\zlib-ng\zlib-ng.vcxproj]
D:\a\zlib-ng\zlib-ng\inffast_tpl.h(308): warning C4244: '-=': conversion from 'unsigned int' to 'uint8_t', possible loss of data [D:\a\zlib-ng\zlib-ng\zlib-ng.vcxproj]
```

--- a/infback.c
+++ b/infback.c
@@ -154,7 +154,7 @@ int32_t Z_EXPORT PREFIX(inflateBack)(PREFIX3(stream) *strm, in_func in, void *in
     unsigned char *put;          /* next output */
     unsigned have, left;         /* available input and output */
     uint64_t hold;               /* bit buffer */
-    uint8_t bits;                /* bits in bit buffer */
+    bits_t bits;                 /* bits in bit buffer */
     unsigned copy;               /* number of stored or match bytes to copy */
     unsigned char *from;         /* where to copy match bytes from */
     code here;                   /* current decoding table entry */

--- a/infback.c
+++ b/infback.c
@@ -154,7 +154,7 @@ int32_t Z_EXPORT PREFIX(inflateBack)(PREFIX3(stream) *strm, in_func in, void *in
     unsigned char *put;          /* next output */
     unsigned have, left;         /* available input and output */
     uint64_t hold;               /* bit buffer */
-    unsigned bits;               /* bits in bit buffer */
+    uint8_t bits;                /* bits in bit buffer */
     unsigned copy;               /* number of stored or match bytes to copy */
     unsigned char *from;         /* where to copy match bytes from */
     code here;                   /* current decoding table entry */

--- a/inffast_tpl.h
+++ b/inffast_tpl.h
@@ -305,7 +305,7 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
     /* return unused bytes (on entry, bits < 8, so in won't go too far back) */
     len = bits >> 3;
     in -= len;
-    bits -= len << 3;
+    bits -= (uint8_t)(len << 3);
     hold &= (UINT64_C(1) << bits) - 1;
 
     /* update state and return */

--- a/inffast_tpl.h
+++ b/inffast_tpl.h
@@ -101,13 +101,13 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
        with (1<<bits)-1 to drop those excess bits so that, on function exit, we
        keep the invariant that (state->hold >> state->bits) == 0.
     */
-    unsigned bits;              /* local strm->bits */
+    uint8_t bits;               /* local strm->bits */
     uint64_t hold;              /* local strm->hold */
     unsigned lmask;             /* mask for first level of length codes */
     unsigned dmask;             /* mask for first level of distance codes */
     code const *lcode;          /* local strm->lencode */
     code const *dcode;          /* local strm->distcode */
-    const code *here;           /* retrieved table entry */
+    code here;                  /* retrieved table entry */
     unsigned op;                /* code bits, operation, extra bits, or */
                                 /*  window position, window bytes to copy */
     unsigned len;               /* match length, unused bytes */
@@ -128,7 +128,7 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
     wnext = state->wnext;
     window = state->window;
     hold = state->hold;
-    bits = state->bits;
+    bits = (uint8_t)state->bits;
     lcode = state->lencode;
     dcode = state->distcode;
     lmask = (1U << state->lenbits) - 1;
@@ -149,40 +149,44 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
        input data or output space */
     do {
         REFILL();
-        here = lcode + (hold & lmask);
-        if (here->op == 0) {
-            *out++ = (unsigned char)(here->val);
-            DROPBITS(here->bits);
-            here = lcode + (hold & lmask);
-            if (here->op == 0) {
-                *out++ = (unsigned char)(here->val);
-                DROPBITS(here->bits);
-                here = lcode + (hold & lmask);
+        here = lcode[hold & lmask];
+        Z_TOUCH(here);
+        if (here.op == 0) {
+            *out++ = (unsigned char)(here.val);
+            DROPBITS(here.bits);
+            here = lcode[hold & lmask];
+            Z_TOUCH(here);
+            if (here.op == 0) {
+                *out++ = (unsigned char)(here.val);
+                DROPBITS(here.bits);
+                here = lcode[hold & lmask];
+                Z_TOUCH(here);
             }
         }
       dolen:
-        DROPBITS(here->bits);
-        op = here->op;
+        DROPBITS(here.bits);
+        op = here.op;
         if (op == 0) {                          /* literal */
-            Tracevv((stderr, here->val >= 0x20 && here->val < 0x7f ?
+            Tracevv((stderr, here.val >= 0x20 && here.val < 0x7f ?
                     "inflate:         literal '%c'\n" :
-                    "inflate:         literal 0x%02x\n", here->val));
-            *out++ = (unsigned char)(here->val);
+                    "inflate:         literal 0x%02x\n", here.val));
+            *out++ = (unsigned char)(here.val);
         } else if (op & 16) {                     /* length base */
-            len = here->val;
+            len = here.val;
             op &= MAX_BITS;                       /* number of extra bits */
             len += BITS(op);
             DROPBITS(op);
             Tracevv((stderr, "inflate:         length %u\n", len));
-            here = dcode + (hold & dmask);
+            here = dcode[hold & dmask];
+            Z_TOUCH(here);
             if (bits < MAX_BITS + MAX_DIST_EXTRA_BITS) {
                 REFILL();
             }
           dodist:
-            DROPBITS(here->bits);
-            op = here->op;
+            DROPBITS(here.bits);
+            op = here.op;
             if (op & 16) {                      /* distance base */
-                dist = here->val;
+                dist = here.val;
                 op &= MAX_BITS;                 /* number of extra bits */
                 dist += BITS(op);
 #ifdef INFLATE_STRICT
@@ -277,14 +281,16 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
                         out = CHUNKMEMSET(out, out - dist, len);
                 }
             } else if ((op & 64) == 0) {          /* 2nd level distance code */
-                here = dcode + here->val + BITS(op);
+                here = dcode[here.val + BITS(op)];
+                Z_TOUCH(here);
                 goto dodist;
             } else {
                 SET_BAD("invalid distance code");
                 break;
             }
         } else if ((op & 64) == 0) {              /* 2nd level length code */
-            here = lcode + here->val + BITS(op);
+            here = lcode[here.val + BITS(op)];
+            Z_TOUCH(here);
             goto dolen;
         } else if (op & 32) {                     /* end-of-block */
             Tracevv((stderr, "inflate:         end of block\n"));

--- a/inffast_tpl.h
+++ b/inffast_tpl.h
@@ -101,7 +101,7 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
        with (1<<bits)-1 to drop those excess bits so that, on function exit, we
        keep the invariant that (state->hold >> state->bits) == 0.
     */
-    uint8_t bits;               /* local strm->bits */
+    bits_t bits;                /* local strm->bits */
     uint64_t hold;              /* local strm->hold */
     unsigned lmask;             /* mask for first level of length codes */
     unsigned dmask;             /* mask for first level of distance codes */
@@ -128,7 +128,7 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
     wnext = state->wnext;
     window = state->window;
     hold = state->hold;
-    bits = (uint8_t)state->bits;
+    bits = (bits_t)state->bits;
     lcode = state->lencode;
     dcode = state->distcode;
     lmask = (1U << state->lenbits) - 1;
@@ -305,7 +305,7 @@ void Z_INTERNAL INFLATE_FAST(PREFIX3(stream) *strm, uint32_t start) {
     /* return unused bytes (on entry, bits < 8, so in won't go too far back) */
     len = bits >> 3;
     in -= len;
-    bits -= (uint8_t)(len << 3);
+    bits -= (bits_t)(len << 3);
     hold &= (UINT64_C(1) << bits) - 1;
 
     /* update state and return */

--- a/inflate.c
+++ b/inflate.c
@@ -479,7 +479,7 @@ int32_t Z_EXPORT PREFIX(inflate)(PREFIX3(stream) *strm, int32_t flush) {
     unsigned char *from;        /* where to copy match bytes from */
     unsigned have, left;        /* available input and output */
     uint64_t hold;              /* bit buffer */
-    unsigned bits;              /* bits in bit buffer */
+    bits_t bits;                /* bits in bit buffer */
     uint32_t in, out;           /* save starting available input and output */
     unsigned copy;              /* number of stored or match bytes to copy */
     code here;                  /* current decoding table entry */

--- a/inflate_p.h
+++ b/inflate_p.h
@@ -80,7 +80,7 @@
         next = strm->next_in; \
         have = strm->avail_in; \
         hold = state->hold; \
-        bits = state->bits; \
+        bits = (uint8_t)state->bits; \
     } while (0)
 
 /* Restore state from registers in inflate() */
@@ -105,7 +105,7 @@
    not enough available input to do that, then return from inflate()/inflateBack(). */
 #define NEEDBITS(n) \
     do { \
-        while (bits < (unsigned)(n)) \
+        while (bits < (uint8_t)(n)) \
             PULLBYTE(); \
     } while (0)
 
@@ -117,7 +117,7 @@
 #define DROPBITS(n) \
     do { \
         hold >>= (n); \
-        bits -= (unsigned)(n); \
+        bits -= (uint8_t)(n); \
     } while (0)
 
 /* Remove zero to seven bits as needed to go to a byte boundary */

--- a/inflate_p.h
+++ b/inflate_p.h
@@ -72,6 +72,13 @@
     } while (0)
 #endif
 
+/* Compiler optimization for bit accumulator on x86 architectures */
+#if defined(__x86_64__) || defined(__i386__)
+typedef uint8_t bits_t;
+#else
+typedef unsigned bits_t;
+#endif
+
 /* Load registers with state in inflate() for speed */
 #define LOAD() \
     do { \
@@ -80,7 +87,7 @@
         next = strm->next_in; \
         have = strm->avail_in; \
         hold = state->hold; \
-        bits = (uint8_t)state->bits; \
+        bits = (bits_t)state->bits; \
     } while (0)
 
 /* Restore state from registers in inflate() */
@@ -105,7 +112,7 @@
    not enough available input to do that, then return from inflate()/inflateBack(). */
 #define NEEDBITS(n) \
     do { \
-        while (bits < (uint8_t)(n)) \
+        while (bits < (bits_t)(n)) \
             PULLBYTE(); \
     } while (0)
 
@@ -117,7 +124,7 @@
 #define DROPBITS(n) \
     do { \
         hold >>= (n); \
-        bits -= (uint8_t)(n); \
+        bits -= (bits_t)(n); \
     } while (0)
 
 /* Remove zero to seven bits as needed to go to a byte boundary */

--- a/zbuild.h
+++ b/zbuild.h
@@ -127,6 +127,15 @@
 /* Ignore unused variable warning */
 #define Z_UNUSED(var) (void)(var)
 
+/* Force the compiler to treat variable as modified. Empty asm statement with a "+r" constraint prevents
+   the compiler from reordering or eliminating loads into the variable. This can help keep critical latency
+   chains in the hot path from being shortened or optimized away. */
+#if (defined(__GNUC__) || defined(__clang__)) && (defined(__x86_64__) || defined(__i386__) || defined(__aarch64__))
+#  define Z_TOUCH(var) __asm__ ("" : "+r"(var))
+#else
+#  define Z_TOUCH(var) (void)(var)
+#endif
+
 #if defined(HAVE_VISIBILITY_INTERNAL)
 #  define Z_INTERNAL __attribute__((visibility ("internal")))
 #elif defined(HAVE_VISIBILITY_HIDDEN)

